File: solutions/46.permutations.js
/*
 * @lc app=leetcode id=46 lang=javascript
 *
 * [46] Permutations
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
  // if (nums.length <= 1) return nums.map((e) => [e]);
  // const res = [];
  // for (let i = 0; i < nums.length; ++i) {
  //   permute(nums.slice(0, i).concat(nums.slice(i + 1))).forEach((list) => {
  //     res.push(list.concat(nums[i]));
  //   });
  // }
  // return res;

  const res = [];
  const dfs = (start) => {
    if (start === nums.length) {
      res.push([...nums]);
      return;
    }
    for (let i = start; i < nums.length; ++i) {
      [nums[i], nums[start]] = [nums[start], nums[i]];
      dfs(start + 1);
      [nums[i], nums[start]] = [nums[start], nums[i]];
    }
  };
  dfs(0);
  return res;
};
// @lc code=end

File: solutions/46.permutations.rb
#
# @lc app=leetcode id=46 lang=ruby
#
# [46] Permutations
#

# @lc code=start
# @param {Integer[]} nums
# @return {Integer[][]}
def permute(nums)
  ans = []
  tmp = []
  visited = []
  dfs = lambda do
    return ans << tmp.dup if tmp.length == nums.length

    (0...nums.length).each do |i|
      next if visited[i]
      visited[i] = true
      tmp << nums[i]
      dfs.call
      tmp.pop
      visited[i] = false
    end
  end

  dfs.call
  ans
end
# @lc code=end

